File: str_helper.cpp
// system includes
#include <map>

// includes
#include "str_helper.h"
#include "generic_protocol/str_helper.h"
#include "basic_objects/str_helper.h"
#include "basic_parser/str_helper.h"

namespace user_management_protocol
{

namespace str_helper
{

using ::basic_parser::str_helper::write;
using ::basic_parser::str_helper::write_t;

// enums

#define TUPLE_VAL_STR(_x_)  _x_,#_x_

std::ostream & write( std::ostream & os, const gender_e r )
{
    typedef gender_e Type;
    static const std::map< Type, std::string > m =
    {
        { Type:: TUPLE_VAL_STR( UNDEF ) },
        { Type:: TUPLE_VAL_STR( MALE ) },
        { Type:: TUPLE_VAL_STR( FEMALE ) },
    };

    auto it = m.find( r );

    static const std::string undef( "undef" );

    if( it != m.end() )
        return write( os, it->second );

    return write( os, undef );
}

// objects

std::ostream & write( std::ostream & os, const UserInfo & r )
{
    os << "(";

    os << " gender="; write( os, r.gender );
    os << " last_name="; write( os, r.last_name );
    os << " first_name="; write( os, r.first_name );
    os << " company_name="; write( os, r.company_name );
    os << " email="; write( os, r.email );
    os << " email_2="; write( os, r.email_2 );
    os << " phone="; write( os, r.phone );
    os << " phone_2="; write( os, r.phone_2 );
    os << " timezone="; write( os, r.timezone );

    os << ")";

    return os;
}

// base messages

std::ostream & write( std::ostream & os, const Request & r )
{
    os << "(";

    // no base class

    os << ")";

    return os;
}

std::ostream & write( std::ostream & os, const BackwardMessage & r )
{
    os << "(";

    // base class
    ::generic_protocol::str_helper::write( os, static_cast<const generic_protocol::BackwardMessage&>( r ) );


    os << ")";

    return os;
}

// messages

std::ostream & write( std::ostream & os, const SetUserInfoRequest & r )
{
    // base class
    str_helper::write( os, static_cast<const Request&>( r ) );

    os << " user_id="; write( os, r.user_id );
    os << " user_info="; write( os, r.user_info );

    return os;
}

std::ostream & write( std::ostream & os, const SetUserInfoResponse & r )
{
    // base class
    str_helper::write( os, static_cast<const BackwardMessage&>( r ) );


    return os;
}

std::ostream & write( std::ostream & os, const GetUserInfoRequest & r )
{
    // base class
    str_helper::write( os, static_cast<const Request&>( r ) );

    os << " user_id="; write( os, r.user_id );

    return os;
}

std::ostream & write( std::ostream & os, const GetUserInfoResponse & r )
{
    // base class
    str_helper::write( os, static_cast<const BackwardMessage&>( r ) );

    os << " user_id="; write( os, r.user_id );
    os << " gender="; write( os, r.gender );
    os << " last_name="; write( os, r.last_name );
    os << " first_name="; write( os, r.first_name );
    os << " company_name="; write( os, r.company_name );
    os << " email="; write( os, r.email );
    os << " email_2="; write( os, r.email_2 );
    os << " phone="; write( os, r.phone );
    os << " phone_2="; write( os, r.phone_2 );
    os << " timezone="; write( os, r.timezone );

    return os;
}

} // namespace str_helper

} // namespace user_management_protocol

File: str_helper.php
<?php

namespace user_management_protocol;


// base include
require_once __DIR__.'/../generic_protocol/str_helper.php';
// includes
require_once __DIR__.'/../basic_objects/str_helper.php';
require_once __DIR__.'/../basic_parser/str_helper.php';

// enums

function to_string__gender_e( $r )
{
    $map = array
    (
        gender_e__UNDEF => 'UNDEF',
        gender_e__MALE => 'MALE',
        gender_e__FEMALE => 'FEMALE',
    );

    if( array_key_exists( $r, $map ) )
    {
        return $map[ $r ];
    }

    return '?';
}

// objects

function to_string__UserInfo( & $r )
{
    $res = "";

    $res .= "(";

    $res .= " gender=" . to_string__gender_e( $r->gender );
    $res .= " last_name=" . \basic_parser\to_string__string( $r->last_name );
    $res .= " first_name=" . \basic_parser\to_string__string( $r->first_name );
    $res .= " company_name=" . \basic_parser\to_string__string( $r->company_name );
    $res .= " email=" . \basic_parser\to_string__string( $r->email );
    $res .= " email_2=" . \basic_parser\to_string__string( $r->email_2 );
    $res .= " phone=" . \basic_parser\to_string__string( $r->phone );
    $res .= " phone_2=" . \basic_parser\to_string__string( $r->phone_2 );
    $res .= " timezone=" . \basic_parser\to_string__string( $r->timezone );

    $res .= ")";

    return $res;
}

// base messages

function to_string__Request( & $r )
{
    $res = "";

    $res .= "(";

    // no base class

    $res .= ")";

    return $res;
}

function to_string__BackwardMessage( & $r )
{
    $res = "";

    $res .= "(";

    // base class
    $res .= \generic_protocol\to_string__BackwardMessage( $r );


    $res .= ")";

    return $res;
}

// messages

function to_string__SetUserInfoRequest( & $r )
{
    $res = "";

    // base class
    $res .= to_string__Request( $r );

    $res .= " user_id=" . \basic_parser\to_string__int( $r->user_id );
    $res .= " user_info=" . to_string__UserInfo( $r->user_info );

    return $res;
}

function to_string__SetUserInfoResponse( & $r )
{
    $res = "";

    // base class
    $res .= to_string__BackwardMessage( $r );


    return $res;
}

function to_string__GetUserInfoRequest( & $r )
{
    $res = "";

    // base class
    $res .= to_string__Request( $r );

    $res .= " user_id=" . \basic_parser\to_string__int( $r->user_id );

    return $res;
}

function to_string__GetUserInfoResponse( & $r )
{
    $res = "";

    // base class
    $res .= to_string__BackwardMessage( $r );

    $res .= " user_id=" . \basic_parser\to_string__int( $r->user_id );
    $res .= " gender=" . to_string__gender_e( $r->gender );
    $res .= " last_name=" . \basic_parser\to_string__string( $r->last_name );
    $res .= " first_name=" . \basic_parser\to_string__string( $r->first_name );
    $res .= " company_name=" . \basic_parser\to_string__string( $r->company_name );
    $res .= " email=" . \basic_parser\to_string__string( $r->email );
    $res .= " email_2=" . \basic_parser\to_string__string( $r->email_2 );
    $res .= " phone=" . \basic_parser\to_string__string( $r->phone );
    $res .= " phone_2=" . \basic_parser\to_string__string( $r->phone_2 );
    $res .= " timezone=" . \basic_parser\to_string__string( $r->timezone );

    return $res;
}

// generic

function to_string( $obj )
{
    $handler_map = array(
        // objects
        'user_management_protocol\UserInfo'         => 'to_string__UserInfo',
        // messages
        'user_management_protocol\SetUserInfoRequest'         => 'to_string__SetUserInfoRequest',
        'user_management_protocol\SetUserInfoResponse'         => 'to_string__SetUserInfoResponse',
        'user_management_protocol\GetUserInfoRequest'         => 'to_string__GetUserInfoRequest',
        'user_management_protocol\GetUserInfoResponse'         => 'to_string__GetUserInfoResponse',
    );

    $type = get_class( $obj );

    if( array_key_exists( $type, $handler_map ) )
    {
        $func = '\\user_management_protocol\\' . $handler_map[ $type ];
        return $func( $obj );
    }

    return \generic_protocol\to_string( $obj );
}

// namespace_end user_management_protocol


?>
